control-plane/agents/src/bin/core/tests/rebuild/mod.rs

```diff
@@ -169,14 +176,118 @@ async fn replica_wait_and_full_rebuild() {
+    tracing::info!({?replicas}, "Available replicas - ");
+    let testrep = replicas.iter().find(|r| r.node != vol_target).unwrap();
+    let testrep_node = &testrep.node;
+    tracing::info!(
+        "Stopping node {testrep_node} having replica {}",
+        testrep.uri.clone()
+    );
+
+    // Stop the container and don't restart.
+    cluster
+        .composer()
+        .stop(&testrep_node.to_string())
+        .await
+        .expect("container stop failure");
+
+    // Check the pool on restarted node shouldn't be online.
+    // This must expire the wait timeout.
+    if is_pool_online(&cluster, testrep, Duration::from_secs(20))
+        .await
+        .is_some()
+    {
+        panic!("Pool is not expected to be Online");
+    }
+
+    // Fetch replicas again.
+    let replicas_now = api_client.replicas_api().get_replicas().await.unwrap();
+    // The replica picked for test must not be online now again.
+    assert!(!replicas_now.iter().any(|r| r.uri == testrep.uri));
+
+    // Get the volume again for validations.
+    let vol = volume_api.get_volume(volid.uuid()).await.unwrap();
+    let volume_state = vol.state;
+    let nexus = volume_state.target.unwrap();
+    // The child must not be in the nexus anymore.
+    let is_removed = !nexus.children.iter().any(|c| c.uri == testrep.uri);
+    assert!(is_removed);
+    // TODO: We can additionally check that the volume state eventually becomes Online,
+    // and doesn't remain degraded once the full rebuild finishes. Currently no support for
+    // getting rebuild info via nexus grpc client NexusOperations.
+    let nexus_client = cluster.grpc_client().nexus();
+    let nexus_id = volume.state.target.unwrap().uuid;
+    wait_nexus_online(&nexus_client, NexusId::from(nexus_id))
+        .await
+        .expect("Rebuild didnt finish in 30 secs");
+}
+
+#[tokio::test]
+async fn rebuild_history_for_full_rebuild() {
+    let cluster = build_cluster(4, 52428800, 0).await; // 50MiB pool size
+
+    let vol_target = cluster.node(0).to_string();
+    let api_client = cluster.rest_v00();
+    let volume_api = api_client.volumes_api();
+    let nexus_client = cluster.grpc_client().nexus();
+
+    let body = CreateVolumeBody::new(VolumePolicy::new(true), 3, 10485760u64, false);
+    let volid = VolumeId::new();
+    let volume = volume_api.put_volume(&volid, body).await.unwrap();
+    let volume = volume_api
+        .put_volume_target(
+            &volume.spec.uuid,
+            PublishVolumeBody::new_all(
+                HashMap::new(),
+                None,
+                vol_target.clone().to_string(),
+                models::VolumeShareProtocol::Nvmf,
+                None,
+                cluster.csi_node(0),
+            ),
+        )
+        .await
+        .expect("Failed to publish volume");
+
+    let volume_state = volume.state.clone();
+    let nexus = volume_state.target.unwrap();
+    let nexus_id = NexusId::from(nexus.uuid);
+    sleep(Duration::from_secs(5));
```

Comment:
What's this sleep for?

Reply:
I have seen that as sson as we publish volume rebuild history was not available for the nexus. Hence thought of sleep here. Probably should check interval for state update and use that for sleep?

Reply:
I'm confused, we're just publishing the nexus here right, so there will be no rebuild history?

Reply:
There will be, we will not have any records in the vector. hence matching if nexus uuid is of newly published nexus and record len is 0.